Add exception for missing system requirements

kamalesh0406 opened the pull request:
## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Describe changes
Briefly describe the changes you have introduced.

I have modified the check_installation function and the Integration class to look for missing system requirements particularly pertaining to the graphviz integration. The new code raises an Exception during zenml init were it mentions the library whose package is missing from the system. I tried to make it raise the error during zenml example run but for that to be done the changes should be done in example cli and it was mentioned in the following issue #266 that an override for check_installation is preferable.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/zenml-io/zenml?pullRequest=281) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/zenml-io/zenml?pullRequest=281) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/zenml-io/zenml?pullRequest=281) it.</sub>

AlexejPenner commented on src/zenml/integrations/integration.py:
```diff
@@ -42,10 +44,19 @@ class Integration(metaclass=IntegrationMeta):
+                result = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True, shell=True)
+                if result.returncode!=0:
+                    raise DoesNotExistException(
+                            f"Unable to find linux/unix installation of {req}"
```
In my opinion this message should be os - agnostic. Maybe also add a sentence on what the user should do to fix this. Maybe a second sentence like: “Please install the package and retry.”

kamalesh0406 replied:
Yes, even I was thinking about this. I will try to use platform to print a more OS-specific message. The only issue is platform.system() prints 'Darwin' for macOS and that might confuse the user at times.

AlexejPenner replied:
Alternativly, you could also just not mention the platform at all. The user already knows his platform and can find the appropriate way to install the package.

kamalesh0406 replied:
Yup,  I will do that.

AlexejPenner reviewed (changes requested):
Overall, the code changes look good to me. I’m away from my machine so I can’t verify functionality right now. I added a small change request in the comments regarding the error message.

Thanks a lot for your contribution Kamalesh☺️

michael-zenml commented:
I like the implementation using the additional list of system requirements! We might consider using `shutil.which(...)` to check if a command exists without actually running it using `subprocess.run(...)`, which would allow us to simplify the system requirements to `["dot"]` (without the -V).

kamalesh0406 commented:
> I like the implementation using the additional dict of system requirements! We might consider using `shutil.which(...)` to check if a command exists without actually running it using `subprocess.run(...)`, which would allow us to simplify the system requirements to `["dot"]` (without the -V).

Thank you for the suggestion, I will change the subprocess to shutil.which .

AlexejPenner commented:
Makes sense that we have failing tests now, as they also have a dependency on check_installation. For the scope of this bugfix it might make sense to hard code the installation of graphviz on the level of our github actions. In the long run we could then maybe have a feature ticket to implement this in a more generic way for all possible future “System requirements” as well. 

Do you need any additional guidance regarding our github actions?

kamalesh0406 commented:
> Makes sense that we have failing tests now, as they also have a dependency on check_installation. For the scope of this bugfix it might make sense to hard code the installation of graphviz on the level of our github actions. In the long run we could then maybe have a feature ticket to implement this in a more generic way for all possible future “System requirements” as well.
> 
> Do you need any additional guidance regarding our github actions?

So the idea is to now ensure that graphviz is installed on the system that the github actions is testing? Please do correct me if I am wrong.

kamalesh0406 commented:
I have modified the error message, subprocess library and the github actions as well in the new commit.

AlexejPenner commented:
> I have modified the error message, subprocess library and the github actions as well in the new commit.

Thanks Kamalesh. I don’t see any changes on the github actions within your commit. I think you may have forgotten to commit or push your changes within the .github folder?

kamalesh0406 commented:
> > I have modified the error message, subprocess library and the github actions as well in the new commit.
> 
> Thanks Kamalesh. I don’t see any changes on the github actions within your commit. I think you may have forgotten to commit or push your changes within the .github folder?

Yup sorry I forgot to push them. I have modified the commit and pushed it again.

AlexejPenner commented on .github/workflows/pull_request.yml:
```diff
@@ -24,6 +24,14 @@ jobs:
         with:
           python-version: ${{ matrix.python-version }}
 
+      - name: Install System Dependencies
```
I believe the correct syntax here would be to name each condition separately:

```yaml
    - name: condition no 1
      if: condition_1=true
      run: script for condition 1
   - name: condition 2
      if: condition_2=true
      run: script for condition 2
```

kamalesh0406 replied:
Okay, that makes sense. I have pushed the changes, sorry for the inconvenience.

kamalesh0406 replied:
The test failed since I used double quotes instead of using single quotes for specifying the system.

AlexejPenner replied:
Thanks for all your efforts. Looks good to me so far 😊

Are you able to see the logs of the Github actions? There seem to be some linting issues. If you have no access I can paste the logs here ☺️

kamalesh0406 replied:
Yup I have modified my code according to the error. :)

htahir1 commented:
@kamalesh0406 In order to fix these linting issues, we have a script in the `scripts` folder called [format.sh](https://github.com/zenml-io/zenml/blob/main/scripts/format.sh). Run it as  follows:

```
cd zenml/
poetry run bash scripts/format.sh
```
or simply:
```
./scripts/format.sh
```

You can also run linting as follows:
```
poetry run bash scripts/lint.sh
poetry run bash scripts/check-spelling.sh
```

And tests as follows:

```
poetry run bash scripts/test-coverage-xml.sh
```

I'll add this to the CONTRIBUTING.md for future reference

AlexejPenner commented on src/zenml/integrations/integration.py:
```diff
@@ -14,9 +14,11 @@
 from typing import Any, Dict, List, Tuple, Type, cast
 
 import pkg_resources
+import shutil
 
 from zenml.integrations.registry import integration_registry
 from zenml.logger import get_logger
+from zenml.exceptions import DoesNotExistException
```
This is the order of imports you would get from running  format.sh that @htahir1 was suggesting. Adding it here to save you some time :)

```suggestion
import shutil
from typing import Any, Dict, List, Tuple, Type, cast

import pkg_resources

from zenml.exceptions import DoesNotExistException
from zenml.integrations.registry import integration_registry
from zenml.logger import get_logger
```

AlexejPenner commented:
@kamalesh0406 Sorry for pushing directly onto your repo, I must have pressed the wrong button, was just gonna add it as a suggestion. I think now there's just a newline missing at the end of the file before we can merge

kamalesh0406 commented:
Okay I will check it and push the changes.

kamalesh0406 commented:
I am not able to recreate the lining issue on my local, can you specify the exact issue? The github actions output is not very clear as well.

AlexejPenner commented:
> I am not able to recreate the linting issue on my local, can you specify the exact issue? The github actions output is not very clear as well.

Locally within your repo you should be able to just run `poetry run bash scripts/format.sh` - this will fix the issue for you. Then you just need to commit and push the file.

AlexejPenner commented:
@kamalesh0406 Looks great, thanks for your contribution :pray:
